service: raft: force initial snapshot transfer in new cluster

kbr-scylla opened the pull request:
When we upgrade a cluster to use Raft, or perform manual Raft recovery procedure (which also creates a fresh group 0 cluster, using the same algorithm as during upgrade), we start with a non-empty group 0 state machine; in particular, the schema tables are non-empty.

In this case we need to ensure that nodes which join group 0 receive the group 0 state. Right now this is not the case. In previous releases, where group 0 consisted only of schema, and schema pulls were also done outside Raft, those nodes received schema through this outside mechanism. In 91f609d065d5b83043a814c1169ea26ffb9ffe61 we disabled schema pulls outside Raft; we're also extending group 0 with other things, like topology-specific state.

To solve this, we force snapshot transfers by setting the initial snapshot index on the first group 0 server to `1` instead of `0`. During replication, Raft will see that the joining servers are behind, triggering snapshot transfer and forcing them to pull group 0 state.

It's unnecessary to do this for cluster which bootstraps with Raft enabled right away but it also doesn't hurt, so we keep the logic simple and don't introduce branches based on that.

Extend Raft upgrade tests with a node bootstrap step at the end to prevent regressions (without this patch, the step would hang - node would never join, waiting for schema).

Fixes: #14066

gleb-cloudius commented on test/topology_raft_disabled/test_raft_upgrade_basic.py:
```diff
@@ -25,7 +25,7 @@
 async def test_raft_upgrade_basic(manager: ManagerClient, random_tables: RandomTables):
     """
-    kbr-: the test takes about 7 seconds in dev mode on my laptop.
+    kbr-: the test takes about 12 seconds in dev mode on my laptop.
```
Will you patch it each time your laptop changes? :)

kbr-scylla replied:
Removed the comment

kbr-scylla commented:
@tgrabiec please review

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1974/

kbr-scylla commented:
> CI state FAILURE - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1974/

two failures...
- [test_topology_schema.test_topology_schema_changes](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1974/testReport/junit/(root)/test_topology_schema/Tests___Unit_Tests___test_topology_schema_changes/) failed because of https://github.com/scylladb/scylladb/issues/14042
- [cdc_test.TestCdc.test_cluster_reduction_with_cdc[Multi_DC_cluster]](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1974/testReport/junit/cdc_test/TestCdc/Tests___Sanity_Tests___test_cluster_reduction_with_cdc_Multi_DC_cluster_/) and the other test failed because dtest detected an ERROR in the leader's log:
```
E           AssertionError: Unexpected errors found: [('node1', ['ERROR 2023-06-21 12:32:48,455 [shard 0] raft - [e09b08a1-1f29-4745-b70a-0dc4f8f729df] Transferring snapshot to 9ff728b4-7285-4735-b05d-2c4705883d87 failed with: seastar::rpc::remote_verb_error (connection is closed)'])]
```

I don't know why this error happened, but the bootstrap succeeded anyway and in general the test finished, I guess that the transfer was retried and it succeeded on the second attempt.
And dtest considers any `ERROR` in the log to indicate a problem by default.

kbr-scylla commented:
@gleb-cloudius any idea why are we getting these spurious `seastar::rpc::remote_verb_error (connection is closed)`?
Maybe it's the connection dropping thing again... We solved it for repair/streaming so that it's not interrupted using `wait_for_normal_state_handled_on_boot`. But it's called after joining group 0. Maybe we should move it to happen even earlier... ugh.

kbr-scylla commented:
>  Maybe it's the connection dropping thing again... We solved it for repair/streaming so that it's not interrupted using wait_for_normal_state_handled_on_boot. But it's called after joining group 0. Maybe we should move it to happen even earlier... ugh.

https://github.com/scylladb/scylladb/issues/12764
https://github.com/scylladb/scylladb/issues/12956
https://github.com/scylladb/scylladb/issues/13302

gleb-cloudius commented:
https://github.com/scylladb/scylladb/pull/14336#issuecomment-1600855823

gleb-cloudius commented:
https://github.com/scylladb/scylladb/issues/12972#issuecomment-1531232286

avikivity commented:
How can such brokenness exist? Doesn't raft force a snapshot transfer? Or did we just "forget" to implement the callback?

kbr-scylla commented:
Raft transfers a snapshot if there's a snapshot. If the log is empty there is no snapshot so nothing is transferred. After upgrade or recovery we would have started with an empty Raft log hence nothing would be transferred.

In this PR we create an "artificial" snapshot by initializing the Raft log at index 1 instead of 0 (0 meaning empty log). That will indeed trigger a snapshot transfer.

avikivity commented:
But a new node pulls non-empty data.

kostja commented on service/raft/raft_group0.cc:
```diff
@@ -420,14 +420,22 @@ future<> raft_group0::join_group0(std::vector<gms::inet_address> seeds, bool as_
+                // Force snapshot transfer from us to subsequently joining servers.
+                // This is important for upgrade and recovery, where the group 0 state machine
+                // (schema tables in particular) is nonempty.
+                // In a fresh cluster this will trigger an empty snapshot transfer which is redundant but correct.
```
I don't think the snapshot cluster in a fresh cluster will be "empty". It will contain mutations from system tables, which are not empty. This was exactly my point during the call. Previously, Scylla fresh cluster setup was not transferring any data. This was important for example when you create a fresh cluster using *heterogeneous* versions. In this case each node would have its own initial state of the system tables. After this change, the leader of the cluster will be pushing its system tables state over to all nodes, regardless of their version. It could be harmless, but it is not asked for.

kbr-scylla replied:
> I don't think the snapshot in a fresh cluster will be "empty". It will contain mutations from system tables, which are not empty.

Right, I'll fix the comment.

> It could be harmless

It **must** be harmless, if not - we have a bug.

kbr-scylla replied:
> Right, I'll fix the comment.

Actually, it's wrong... The snapshot will be empty.

What do you mean by "state of system tables"? I thought you're talking about schema and topology tables. If those are nonempty, then Raft log is nonempty - we modify them by committing commands after all.

We don't send any internal state like `system.local`. That would be unsafe of course.

So, in a fresh cluster the snapshot **will** be empty but the Raft log won't be empty because of initially created distributed tables.

> Imagine a fresh cluster containing two nodes, 5.2 and 5.3. 5.3 snapshot will contain 5.3 mutations. They will be sent over to 5.2 node is 5.3 is the leader. These mutations are not guaranteed to be harmless on 5.2.

In the same way any commands committed by the 5.3 node may not be safe. 

This is an orthogonal problem to snapshotting. It's the problem of cluster features and implementing a correct upgrade procedure in general. A 5.2 node should not be able to join 5.3 cluster if there's an incompatible feature.

kostja replied:
You're right that there is no practical difference today and the situation I described should be handled by the feature compatibility check.

I still cringe when I reason about this code, however, and here's why.

This patch establishes and documents a new contract - which was until now implicit:
- snapshot index 0 means the snapshot is empty and Raft leader doesn't need to transfer a snapshot
- snapshot index 1 means the snapshot is non-empty and Raft leader needs to transfer a snapshot to every new follower.

And then you misuse this contract in this very patch. A new node has an empty raft snapshot, and we used snapshot index 0 for it until now. Yet after this patch, whether or not a node is fresh, we will use snapshot index 1. Of course transferring an empty snasphot is harmless (in theory). But if it is harmless, what use case are we supporting by allowing snapshot index 0? Maybe we should require snapshot index 1 in all cases?

I think the use of index 1 in the code blurs the very protocol the code establishes.

kostja replied:
Given the patch is fixing a CI issue, perhaps it should be approved. But perhaps I at least was able to make my point clear.

kbr-scylla replied:
> snapshot index 1 means the snapshot is non-empty and Raft leader needs to transfer a snapshot to every new follower.

> And then you misuse this contract in this very patch.

None of the comments said "snapshot index 1 means that the snapshot is non-empty". The comments are only saying that snapshot index 1 will enforce snapshot transfer, which is **necessary** when your snapshot is non-empty, but it doesn't imply that your snapshot is non-empty. And we require nothing about empty snapshots in this case, it's fine to transfer them, it's fine not to transfer them.

gleb-cloudius replied:
> None of the comments said "snapshot index 1 means that the snapshot is non-empty".

In fact it can be empty since the command may have been "dummy".

kostja replied:
Alright, but what point does it make to allow snapshot index 0 then? We ourselves are not using it.

kbr-scylla replied:
We are setting snapshot index to 1 only on the first server in group 0. Second, third etc. must have it set to 0. Otherwise Raft would assume that log is in sync and not send a snapshot.

kbr-scylla replied:
But for the first server indeed there is no point to ever set it to 0. Maybe as a micro-optimization if you know that your snapshot is empty, but not worth the brain cycles.

kostja replied:
// Note: there are two sensible options for the initial snapshot
    // used by the first server in a Raft cluster:
    // - if the initial snapshot's index is 0, Raft will assume that
    //   the state machine is empty,
    // - if the initial snapshot's index is 1, Raft will have to perform
    //   a snapshot transfer before replicating commands to any server
    //   which joins with empty Raft log (in particular with  snapshot index 0).
    // Thus, if you want to create a Raft cluster with a non-empty state
    // machine, so that joining servers always receive a snapshot,
    // you should set the initial snapshot index on the first server to 1
    // and on all subsequently joining servers to 0.
    
    
    You say here specifically: there is no point in having snapshot index 0 for the leader. 
    You say in the comment: there are two sensible options for the first server. 
    Looks like option of index 0 is not sensible after all?

kostja replied:
Ideally, let's ban snapshot index 0 for the first server. I understand detecting the first server on raft level is impossible. But then again, what is our stock suggestion? After all, for raft groups which don't require a pre-defined initialization order, it makes sense to set snapshot index 0 on all servers? The implicitness of the contract is what frustrates me.

kbr-scylla replied:
> You say in the comment: there are two sensible options for the first server. 
Looks like option of index 0 is not sensible after all?

It is sensible as a micro-optimization when you have an empty snapshot.

> But then again, what is our stock suggestion? After all, for raft groups which don't require a pre-defined initialization order, it makes sense to set snapshot index 0 on all servers? The implicitness of the contract is what frustrates me.

The stock suggestion would be:
- for all servers present in the initial configuration, set snapshot index to 1. (And make sure that they have the same initial state machine state.)
- for all other servers (i.e. those which will join through a configuration change), set snapshot index to 0.

kostja commented on service/raft/raft_sys_table_storage.cc:
```diff
@@ -301,8 +301,9 @@ future<> raft_sys_table_storage::execute_with_linearization_point(std::function<
 
-future<> raft_sys_table_storage::bootstrap(raft::configuration initial_configuation) {
-    raft::snapshot_descriptor snapshot;
+future<> raft_sys_table_storage::bootstrap(raft::configuration initial_configuation, bool nontrivial_snapshot) {
```
Following my earlier comment, I would avoid sending a snapshot when bootstrapping a fresh, empty cluster, I think we should pass group0_upgrade_state instead here, or, if we stick to a boolean flag, rename bool nontrivial_snapshot to bool in_recovery.

kostja commented on test/topology_raft_disabled/test_raft_upgrade_basic.py:
```diff
@@ -53,3 +50,8 @@ async def test_raft_upgrade_basic(manager: ManagerClient, random_tables: RandomT
+
+    logging.info("Booting new node")
+    await manager.server_add(config={
+        'consistent_cluster_management': True
```
At least a comment is missing that we are actually overriding the pre-existent setting to False. consistent_cluster_management: True is set by default in the current master, so it is not obvious why we have to add the very same setting here.

kbr-scylla replied:
The suite is called `topology_raft_disabled`. The whole **point** of the test is to have Raft disabled by default, after all it's an upgrade test. Adding a comment like "the default is False" is completely redundant. I can add it but there's no point really.

gleb-cloudius commented:
> But a new node pulls non-empty data.

In raft new nodes pull nothing. Everything is pushed by a leader and if a leader thinks that a node's state machine is up-to-date it will not push anything.

kbr-scylla commented:
I sent https://github.com/scylladb/scylladb/pull/14354 to fix the cause of the CI failure described here: https://github.com/scylladb/scylladb/pull/14336#issuecomment-1600850848

After this, we should no longer be getting snapshot transfer errors so dtests won't complain.
The gossiper notification issue should also be fixed by the above PR.

tgrabiec commented:
Is this blocked by https://github.com/scylladb/scylladb/pull/14354 ?

kbr-scylla commented:
> Is this blocked by https://github.com/scylladb/scylladb/pull/14354 ?

Yes, I'll follow-up

kbr-scylla commented:
Rebased and modified the comment in raft.hh to:
```
    // If you want to create a Raft cluster with a non-empty state
    // machine, so that joining servers always receive a snapshot,
    // you should:
    // - make sure that members of the initial configuration have
    //   the same state machine state,
    // - set the initial snapshot index on members of the initial
    //   configuration to 1,
    // - set the initial snapshot index on all subsequently joining
    //   servers to 0.
    // This also works if you start with an empty state machine,
    // so consider it as the go-to default.
```

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2155/

kbr-scylla commented:
@tgrabiec please merge
